Adding saving and config GUIs to the @Config based configuration System

thilokru opened the pull request:
Currently, the annotation based configuration system provides no easy way to get access to the necessary objects to create a configuration GUI. Also it is currently not possible to save changes, which were made programatically to the configuration variables, to disk.

This PR changes the following:
It renames the `ConfigManager.load()` method to `ConfigManager.sync()` as it is capable of saving changes made to the Configuration object to the config file.
The usage of this method to write to disk is needed, as changes to the Configuration object are now possible with the now creatable configuration GUIs, aswell as with the new `ConfigManager.updateConfig()` method, which applies changes made to the variables the config is saved in to the Configuration object.

The class `ConfigElement` was changed to provide Methods to get an object for Configuration objects managed by the ConfigManager.

The already existing test mod ConfigTest was changed to showcase the new methods used to create a ConfigGui and to showcase the new saving method in `init()`.

This PR closes Issue #3729

Commit:
Renamed ConfigManager.load() to .sync() as it can also write to FS.

Commit:
Adds easy way to get necessary data for configuration GUIs

Commit:
Changes to allow the creation of the config GUI and to the test mod.

Commit:
Removed some unneccessary testcode.

Commit:
New method to update config from vars.

Commit:
New Method updateConfig, which loads the variables into the Config

Calling updateConfig() and then sync() now causes the variables to be
written to disc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3735) <br/>All committers have signed the CLA.

liach commented on src/main/java/net/minecraftforge/common/config/ConfigElement.java:
```diff
@@ -360,4 +360,40 @@ public Object getMaxValue()
+    
+    public static ConfigElement from(String modid)
+    {
+        return from(modid, null);
```
Mark `@Nullable`s if you are going to use `null`

LexManos replied:
why would this ever be called with a null name?  They should ateleast know what config they are trying to read, or pull in a Class and deal with it that way.

thilokru replied:
The idea was, that in the @Config annotation, you don't have to specify a name if you want to work with the default config file. But as I am going to change that (I didn't think about just passing the Class with the annotation) this will be obsolete.

LexManos replied:
You dont have to specify the name, it defaults to modid.

liach commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -161,10 +161,132 @@ public static void load(String modid, Config.Type type)
     //                    INTERNAL
     // =======================================================
+    static Configuration getConfiguration(String modid, String name) {
+        if(name == null || name.isEmpty())
```
`Strings.isNullOrEmpty(name)`

liach commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -161,10 +161,132 @@ public static void load(String modid, Config.Type type)
+    {
+        for (Field f : cls.getDeclaredFields())
+        {
+            if (!Modifier.isPublic(f.getModifiers()))
```
Why must they be public? A feature like injection may be good.

thilokru replied:
What do you mean with 'feature like injection'? I thought the user would change the variables and would call this method to update the config. Therefore it must be accessible by the mod and must be public.

liach replied:
You should allow mods to use only getters or setters to change variables instead of exposing the field to all the dangers

thilokru replied:
First, I think I'm only exposing a method similar to what already is exposed (see code in sync()).
Second, isn't the point of all of this to get rid of getters and setters? The @Config annotation allows for mods to specify a class as config, which Forge then fills when sync is called. The point of this method (updateConfig()) is to notify forge, that the variables of the mod have been changed by other code and the internal Configuration object needs updating.

LexManos replied:
Thew point isnt to get rid of getters/setters, the point is to get rid of nasty interaction with the Old-as-crap Configuration system we have right now and move to a more abstract system we can deal with in the backend ourselves.

IF we want to get fancy we can define getters and setters ourselves using a transformer so that it just 'magically' tracks the changes, but that's out of the scope of the current state.

liach reviewed (changes requested):
The pr is rather large. Although there are some minor issues, overall this is a great work!

Commit:
Changes to make use of Strings.isNullOrEmpty() and @Nullable

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigElement.java:
```diff
@@ -360,4 +362,40 @@ public Object getMaxValue()
+     * @param categoryName The category name stated in the {@code @Config} annotation.
+     * @return A ConfigElement based on the described category.
+     */
+    public static ConfigElement from(String modid, @Nullable String name, @Nullable String categoryName)
```
if all of this is stated in the @Config annotation why not just pull int he class with the @config annotation?

thilokru replied:
That is a good idea. I will implement this.

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -117,7 +118,7 @@ public static void loadData(ASMDataTable data)
 
-    public static void load(String modid, Config.Type type)
+    public static void sync(String modid, Config.Type type)
     {
```
The fundemental concept of changing this load to a sync that is called multiple times is bad because we're recreating the config object everytime. Which destroies the defaults and anything like that.

We need a load, which is called once by our config management system which loads the comments, loads the defaults and ranges, then we need one where it just syncs it between Class->Config for when modder change the class directly.
And then we need one that syncs Config->Class for when the GUI edits things.

LexManos replied:
Note, it may be worth making the internal function that walks the config class to build the tree private and pass in the 'direction' it's going. 
```
private static void walkConfig(String modid, String category, Configuration cfg, Class<?> ftype, Field f, Object instance, boolean isCreation, boolean toObject){
...
  if (toObject) set(field, cfg.get())
  else              cfg.set(name, getValue(field))
  if (isCreate)
    cfg.getProp(name).setDefault().setRange().setComment()...
...
}
```
This way you don't have to duplicate the bulk of the function which is walking the class tree.

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -161,10 +162,132 @@ public static void load(String modid, Config.Type type)
+    
+    public static void updateConfig(String modid, Config.Type type)
+    {
+        FMLLog.fine("Attempting to update Configuration objects of mod %s for type %s", modid, type);
```
Again, who is calling this, and why are they updating all of the specific of the type at once?

LexManos commented on src/main/java/net/minecraftforge/common/config/TypeAdapters.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Minecraft Forge
+a * Minecraft Forge
  * Copyright (c) 2016.
```
?

thilokru replied:
? Ehm...

LexManos commented on src/test/java/net/minecraftforge/debug/ConfigTest.java:
```diff
@@ -99,4 +133,48 @@ public SubCat(String value)
         }
     }
+
+    public static class ConfigurationGui extends GuiConfig
```
Honestly can we not supply a generic GUI Class implementation that does all the element collection for them? We should be able to respect the hierarchy of the class with @Config. point being is that shouldnt have to deal with even setting the guiFactory or the ConfigChangedEvent handler. It should jsut supply it for them. Unless they override them manually.

thilokru replied:
I'm currently looking into it, but it seems to require a breaking change. If I could replace the method IModGuiFactory.mainConfigGuiClass() with something that returns an already initialized GuiScreen, then this would be a good Idea. Currently you still need to implement an own class providing the arguments for the GuiConfig constructor, just so you can use IModGuiFactory.mainConfigGuiClass().

thilokru replied:
@LexManos I've now implemented an constructor for GuiConfig accepting the annotated classes as parameter. But due to the limmitation of IModGuiFactory.mainConfigGuiClass(), the modder still needs to create his own, but rather small, class to return in this method. I've also added documentation to the new and changed methods. Would you mind having another look at it?

LexManos replied:
The only place mainConfigGuiClass is used is in GuiModList:285 and 418 it would be possible to edit FMLClietnHandler.getGuiFactoryFor to return a implementation that we could manually detect and initalize properly, or, we could deprecate IModGuiFactory.mainConfigGuiClass entirely and change it to a fatory method instead of doing reflection {cpw designed it reflectively for unknown reasons...}

The simplest solution would be hacky, but it would be simple, GuiModList:285:
if (guiFactory.mainConfigGuiClass() == OurDefault.class) getConstructor(GuiScreen.class, String.class).newInstance(this, modid)

thilokru replied:
In the most recent commit, I changed it to a factory method as suggested. This seems better than the hacky way.

LexManos replied:
except you break binary compatibility when you do that, you need to keep but deprecate the old one and wrap it in a check.

thilokru replied:
Here is my proposed changes to restore binary compatibility, but I'm not sure if it would work, nor do I know how to test it:
[Gist - Changes](https://gist.github.com/thilokru/8f2bcc20c3c88cf357bce678619bf223)
Would it work? If not, why? What needs to be changed?

LexManos replied:
That would not work because mainConfigGui would throw an AbstractMethodException.
Because the old implementors of this interface would not have that method existing at all.
What you could do is catch that exception and fall back silently to the old method.
That's the simplest way to do it. Or you could detect that the IModGuiFactory is missing that method {reflect the class at registration time to see if the method exists, or just try calling it and seeing if it AbtractErrors} and wrap it in another bounder class that implements mainConfigGui

Also, I would probably modify the name of that method to createConfigGui or something. A bit more specific on what the method should do.

thilokru replied:
Ok, I changed the method name and fixed the documentation.

thilokru commented:
Thanks for the feedback so far, it is appreciated. I will do my best to implement it, but it will take some time.

Commit:
Changes ConfigElement.from method to accept @Config-Class

Commit:
Merged updateConfig() and sync(), removing redundant code

As in updateConfig() it is detectable, if a config is loaded for the
first time, I merged updateConfig and sync(), and differentiate their
behavior when working with the individual fields.

Commit:
Rewinds to working subcategory system and bugfix in ConfigCategory

When saving the configuration, ConfigCategory did not reset a properties
changed state.

Commit:
Working, simplified config GUIs, changes to @LangKey

First, with this commit, there is a constructor for GuiConfig, which
accepts an array of classes, which have a @Config annotation
Second, with the modifications to @LangKey, the config classes can have
language keys, making them human readable in the config gui.

thilokru commented:
Ok, these commits should address all the requested changes. I have merged sync() und updateConfig() into one Method, now sync(), which determines itself, wether the config needs to be loaded or synced.
The from() method in ConfigElement now accepts the class which has been annotated with @Config. Also, I changed @Config.LangKey to be used on classes, so that the from() method can get a custom language key. I also fixed a minor bug in the Property class. Now, when it is saved, the changed status is reset.

[Edit] Fixed spelling in comment

Commit:
Removes random character.

Commit:
Adds documentation, minor changes to ConfigGui for single classes

Commit:
Changes IModGuiFactory to get initialized config screens

This simplifies the implementation for modders - they can, if they use
the @Config configuration method, return an instance of GuiConfig. There
is a constructor in place for this.

Commit:
Restores backwards and binary compatibility

Commit:
Changes method name to createConfigGui, fixes spelling in documentation

thilokru commented:
Ok, the newest changes should address the comments since the last message. I need another review, as I see no further problems (your opinion may vary), so I can't address them.

bs2609 commented:
@mezz, mind reviewing this when you get time?

mezz commented:
I think this one is better reviewed by Lex directly since it really depends on the design he has in mind.

LexManos commented on src/main/java/net/minecraftforge/fml/client/GuiModList.java:
```diff
@@ -415,8 +423,18 @@ else if (disableable != Disableable.YES)
-            configModButton.enabled = guiFactory != null && guiFactory.mainConfigGui(this) != null;
-
+            configModButton.enabled = false;
+            if(guiFactory != null)
```
This is a concern, do we want things to initalize their screens just to check if they HAVE a screen? that needs to be fixed

thilokru replied:
I don't like making the modder implement yet another method, but I get your point. It may improve perfomance a little bit it and makes it easier to understand for modders. I implemented this in e4b6be4.

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -117,7 +118,24 @@ public static void loadData(ASMDataTable data)
         }
     }
 
-    public static void load(String modid, Config.Type type)
```
This breaks bin compat, if you're renaming the method, toss a bouncer in with a //TODO remove comment

thilokru replied:
This is fixed with commit 782bac8

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -202,17 +230,52 @@ private static void createConfig(String modid, String category, Configuration cf
+                {
+                    set(instance, f, propValue);
+                }
+                return;
```
This is jut a nit-picky code style thing, but if (cond) not if(cond) and for one line if statements no need for braces.

thilokru replied:
I think I should have fixed this with 459f885.

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -202,17 +230,52 @@ private static void createConfig(String modid, String category, Configuration cf
+            {
+                prop = adapter.getProp(cfg, category, f, instance, comment);
+                set(instance, f, adapter.getValue(prop));
+            }
```
Also, cant this all be simplified down to the same prop = cfg.getCat().getProp(name), remove the returns, and wrap the code at the end of the function in a if (loading) setting the comment? because the comment is really the only difference between the two sides of this if statement.

thilokru replied:
This should be fixed with commit 429e6e0

LexManos commented on src/test/java/net/minecraftforge/debug/ConfigTest.java:
```diff
@@ -1,21 +1,50 @@
-@Mod(modid = ConfigTest.MODID, name = "ConfigTest", version = "1.0", acceptableRemoteVersions = "*")
-public class ConfigTest
+@Mod(modid = ConfigTest.MODID, name = "ConfigTest", version = "1.0", acceptableRemoteVersions = "*", guiFactory = ConfigTest.GUI_FACTORY)
+public class ConfigTest implements IModGuiFactory
```
again the goal is that none of this is needed. FMLClientHandler.instance().getGuiFactoryFor(modid) should be able to return a custom instanceof IModGuiFactory that creates the proper GuiConfig for the mod.

thilokru replied:
OK. I'd then add a property to @Config to provide mods a mean to disable the config gui for certain properties, like `@Config(...., guiEnabled = false)` where guiEnabled defaults to true.

LexManos replied:
why? I can understand needing the option to say if it require a restart of MC or not, but why disable flag? The disable flag is them adding a custom guiFactory to their @Mod

thilokru replied:
The flags for MC restart and world restart are added with commit 429e6e0 .
The default config gui is added with 3814cab .

Commit:
Should restore binary compatibility in ConfigManager

Commit:
Rewrites sync(), TypeAdapters and adds map test to test mod.

The code for syncing maps got massive, but it now allows users to add
new keys to the config files, which then will be available to the mod.

Also, the new annotations @RequiresMcRestart and @RequiresWorldRestart
allow the modder to specify, that change to a property a category
requires the specified restart.

Commit:
Fixes language key

Commit:
Finalizes default ConfigGui.

Now the entries are sorted alphabetically.
Also adds a test for the new @RequiresMcRestart annotation.

thilokru commented:
Added features:
+ @RequiresMcRestart and @RequiresWorldRestart
+ Fully working snyc() and massive changes to TypeAdapters
+ DefaultGuiModFactory to provide a default, structured, alphabeticaly sorted configuration gui

Not added after consideration:
- I was asked to not initalize a new GuiConfig every time I want to check if one is available. As this is realy rare (only when the user switches Mods in GuiModList) and Guis are not that expensive to create, I decided against adding an additional method in the Interface IModGuiFactory.

[Edit]
The new commit should fix compiler issues (sadly no Lambda expressions in Java 1.6) and renames one of the new classes for consitstency.
[\Edit]

Commit:
Removes Lambda-expressions (-> Java 1.6), name change for consistency

Commit:
Adds back in an exception in the part of sync() for maps.

Commit:
New hasConigGui() and documentation

First, in IModGuiFactory, the method hasConfigGui() was added, so that
mods may specify if they have a Gui, instead of initializing one to
check if it's non-null.

Second, some documentation was added to IModGuiFactory to make modders
aware of the DefaultGuiFactory class, which provides a default
implementation of IModGuiFactory.

Commit:
Fixes if () and for () formats, removes some {}

Commit:
Introduces field wrappers for generalized field access.

Commit:
Reimplementing syncing with new object oriented system

+IFieldWrapper for generalized fiel access.
+Implementations for Enums, Primitives and Maps in FieldWrapper
+Changes to ITypeAdapter to provide some more useful information.

Commit:
Restoring load on startup, fixes bug in maps.

Commit:
Renaming and cleanup.

Commit:
Format fixes (if and for brackets once again).

thilokru commented:
I was unhappy with the big monolithic method syncField() became. It was horrible to debug and fix errors. Therefore I decided to rewrite that part of the ConfigManager from scratch. It now is a object orriented, flexible system to write variables to disk. It should be easy to add new supported field types, simply by modifying FieldWrapper.get() and adding an appropriate wrapper.

In the code, I found a comment asking wether List-Support should be added. 
My question are:
- Is support for Lists wanted?
- I tried and checked if I reimplemented all old features. Are there any missing?
- What else must or should be changed?

Feedback is appreciated.

LexManos commented:
Revert that last change.
I have specifically opted out of going that route because it exposes an API to modders that I do not wish to do, The ability for them to hack in their own custom types. This is meant to be a hardlist of types not supporting any extension from modders.

thilokru commented:
@LexManos I get your point. You are afraid modders might mess up config files. But this still is a hardcoded list of supported field types. I might have not made me clear: While it is flexible and object oriented, the field types are still hard coded. I'd suggest you taking a look at FieldWrapper.get(). There is no "registry" for field types. Just a bunch of if()-statements. It is as hard to hack in as the old system, but it now is generalized and easier to upgrade in the future.

Edit:
There is, at least I think, no real API being exposed. While modders can use the field wrappers, they don't have access to the configuration objects and therefore shouldn't be able to mess anything up.

LexManos commented on src/test/java/net/minecraftforge/debug/ConfigTest.java:
```diff
@@ -92,11 +127,34 @@ public void preInit(FMLPreInitializationEvent event) {
         {
             @Name("i_say")
-            public static String value;
+            public String value;
```
What?

LexManos commented:
Reading through it, i'd still like it reversed as there isnt really a reason to move it out, and build another abstraction layer on internal stuff.
However, i think im gunna pull this and just clean things up myself when the time comes.
Notably the fact that all of your adaptors can be 1/2 the size with a single helper class.

thilokru commented:
Without this change this would no longer be synced, because it is within
an object. Static fields would stay the same across objects.


Am 01.04.2017 um 23:18 schrieb LexManos:
>
> *@LexManos* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/test/java/net/minecraftforge/debug/ConfigTest.java
> <https://github.com/MinecraftForge/MinecraftForge/pull/3735#discussion_r109294497>:
>
> > @@ -92,11 +127,34 @@ public void preInit(FMLPreInitializationEvent event) {
>          public static class SubCat
>          {
>              @Name("i_say")
> -            public static String value;
> +            public String value;
>
> What?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MinecraftForge/MinecraftForge/pull/3735#pullrequestreview-30391476>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALRnzKPeDs4GzPwrulzzh0VKlf1vlhTFks5rrr9CgaJpZM4MNADM>.
>
